### DISCUSSION TITLE
Is there a way to implement a 3D render that work on most desktop OS?

### DISCUSSION THREAD

**Discussion by sn4k3** [AvaloniaUI/Avalonia#5477] [Q&A]
I need to render a 3D model made from a stack of 2D images.
The 3D scene would be capable of the basic tasks like zoom, pan, rotate.
So far i can't find much documentation or examples of how to implement 3D renders with Avalonia, but i would like to find one that works the same for Windows, Mac and Linux. Maybe openGL?
What can i use and anyone have usable samples?

**Comment by kekekeks** [comment]
There is built-in support for OpenGL(ES) via `OpenGlControlBase`, but there are few caveats:
1) On Windows you need to add `.With(new Win32PlatformOptions {UseWgl = true}`
2) For full portability your code needs to support both OpenGL 3.2+ and OpenGLES 2.0+

**Reply by sn4k3** [reply]
ok, I'm currently using those:

```C#
.With(new Win32PlatformOptions { AllowEglInitialization = true})
.With(new X11PlatformOptions { UseGpu = true, UseEGL = true })
.With(new MacOSPlatformOptions { ShowInDock = true })
.With(new AvaloniaNativePlatformOptions { UseGpu = true })
```

Will UseWgl collide or cause any bad when using the Egl?
I found the OpenGL sample on Control Galarey. Is there any more samples i can look into?

**Reply by kekekeks** [reply]
1) It will take priority over `AllowEglInitialization` and use WGL instead of ANGLE
2) Do NOT use EGL on X11 unless you absolutely need to. It's there for embedded systems with broken dysfunctional proprietary blobs that don't support GLX.

**Reply by sn4k3** [reply]
Ok i have enabled the, `UseWgl = true` this cause my app to start up slow and less fluid, also as soon i open a file it crash without any other change and visual studio don't cath the problem, it just quit the main window. Even if i disable EGL same still happen. I'm using `public override void Render(DrawingContext context)`  on a control and some Skia image manipulation can that be the cause or do i need extra dependencies?

My appbuilder:

```C#
public static AppBuilder BuildAvaloniaApp()
            => AppBuilder.Configure<App>()
                .UsePlatformDetect()
                .With(new Win32PlatformOptions { AllowEglInitialization = true, UseWgl = true})
                .With(new X11PlatformOptions { UseGpu = true, UseEGL = true })
                .With(new MacOSPlatformOptions { ShowInDock = true })
                .With(new AvaloniaNativePlatformOptions { UseGpu = true })
                .UseSkia()
                .LogToTrace();
```

PS: I'm on Windows 10 x64 last updates

**EDIT 1:**  I installed Avalonia.OpenGL don't know if required but still crash after a file load, this time Visual Studio returned this message: 

> The program '[22460] UVtools.exe' has exited with code -1073741819 (0xc0000005) 'Access violation'.
> The program '[22460] UVtools.exe: Program Trace' has exited with code 0 (0x0).

**EDIT 2:** I tried even a simple task, i open my benchmark window, once i click on ComboBox control, it shows the list and then freezes and crash in a second with same violation error
![image](https://user-images.githubusercontent.com/113281/107862744-b5b16500-6e46-11eb-89a4-6b176cf0af2b.png)
